helpers/embeds.js

```diff
@@ -0,0 +1,78 @@
+import { findComposedAncestor } from '../helpers/dom.js';
+import { html } from 'lit';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		for await (const childNode of this.#childNodes) {
+			await childNode.renderNode(options);
+		}
+
+		if (this.#node.id === '3') {
+			console.log('wee');
+		}
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) this.#node.replaceWith(embed);
+	}
+
+}
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	/* NOT IMPLEMENTED */
+}
+
+export async function renderEmbeds(node, options) {
+	const elems = [...node.querySelectorAll('div[data-d2l-embed-type], span[data-d2l-embed-type]')];
+	if (elems.length === 0) return;
+
+	const renderNodes = new Map();
+	const rootNodes = [];
+	const constructRenderNode = elem => {
+		if (renderNodes.has(elem)) return;
+
+		const renderNode = new RenderNode(elem);
+		renderNodes.set(elem, renderNode);
+
+		const nearestEmbedAncestor = findComposedAncestor(elem, ancestor => ancestor !== elem && elems.includes(ancestor));
+		if (!nearestEmbedAncestor) {
+			rootNodes.push(renderNode);
+			return;
+		}
+
+		renderNodes.get(nearestEmbedAncestor).addChildNode(renderNode);
+	};
+
+	elems.forEach(constructRenderNode);
+	rootNodes.forEach(async rootNode => await rootNode.renderNode(options));
```

Comment:
It should always be safe to let the root nodes process in parallel. Though I should change this to a `map` and wrap it in a `Promise.all` to make sure we know all renderers have run before the MathJax and Prism renderers kick in. Right now the `await` in the html-block code won't do much here.